app/slashcommands/command_remote.go

```diff
@@ -0,0 +1,210 @@
+}
+
+func (rp *RemoteProvider) GetAutoCompleteListItems(a *app.App, commandArgs *model.CommandArgs, arg *model.AutocompleteArg, parsed, toBeParsed string) ([]model.AutocompleteListItem, error) {
+	if !a.HasPermissionTo(commandArgs.UserId, model.PERMISSION_MANAGE_SHARED_CHANNELS) {
+		return nil, errors.New("You require `manage_shared_channels` permission to manage remote clusters.")
+	}
+
+	if arg.Name == "remoteId" && strings.Contains(parsed, " remove ") {
+		return getRemoteClusterAutocompleteListItems(a, true)
+	}
+
+	return nil, fmt.Errorf("`%s` not a dynamic argument", arg.Name)
+}
+
+func (rp *RemoteProvider) doAdd(a *app.App, args *model.CommandArgs, margs map[string]string) *model.CommandResponse {
+	name, ok := margs["name"]
+	if !ok {
+		return responsef("Missing `name`")
+	}
+
+	host, ok := margs["host"]
+	if !ok {
+		return responsef("Missing `host`")
+	}
+
+	port, ok := margs["port"]
+	if !ok {
+		port = "8065"
+	}
+	iport, err := strconv.ParseInt(port, 10, 32)
+	if err != nil {
+		return responsef("Invalid `port`: %v", err)
+	}
+
+	rc := &model.RemoteCluster{
+		ClusterName: name,
+		Hostname:    host,
+		Port:        int32(iport),
+	}
+
+	rcSaved, err := a.AddRemoteCluster(rc)
+	if err != nil {
+		responsef("Could not add remote cluster: %v", err)
+	}
+	return responsef("##### Remote cluster added.\nName: %s\nHost: %s\nPort: %d\nToken: %s", rcSaved.ClusterName, rc.Hostname, rc.Port, rcSaved.Token)
+}
+
+func (rp *RemoteProvider) doRemove(a *app.App, args *model.CommandArgs, margs map[string]string) *model.CommandResponse {
+	id, ok := margs["remoteId"]
+	if !ok {
+		return responsef("Missing `remoteId`")
+	}
+
+	deleted, err := a.DeleteRemoteCluster(id)
+	if err != nil {
+		responsef("Could not remove remote cluster: %v", err)
+	}
+
+	result := "removed"
+	if !deleted {
+		result = "**NOT FOUND**"
+	}
+	return responsef("##### Remote cluster %s %s.", id, result)
+}
+
+func (rp *RemoteProvider) doStatus(a *app.App, args *model.CommandArgs, margs map[string]string) *model.CommandResponse {
+	list, err := a.GetAllRemoteClusters(true)
+	if err != nil {
+		responsef("Could not fetch remote clusters: %v", err)
+	}
+
+	if len(list) == 0 {
+		return responsef("** No remote clusters found. **")
+	}
+
+	var sb strings.Builder
+	fmt.Fprintf(&sb, "| Name | Host | Token | Id   | Online |\n")
+	fmt.Fprintf(&sb, "| ---- | ---- | ----  | ---- | ----   |\n")
+
+	for _, rc := range list {
+		online := ":white_check_mark:"
+		if !isOnline(rc.LastPingAt) {
+			online = ":skull_and_crossbones:"
+		}
+
+		fmt.Fprintf(&sb, "| %s | %s:%d | %s | %s | %s |\n", rc.ClusterName, rc.Hostname, rc.Port, rc.Token, rc.Id, online)
+	}
+	return responsef(sb.String())
+}
+
+func isOnline(lastPing int64) bool {
+	return lastPing > model.GetMillis()-model.RemoteOfflineAfterMillis
+}
+
+func getRemoteClusterAutocompleteListItems(a *app.App, incOffline bool) ([]model.AutocompleteListItem, error) {
+	var list []model.AutocompleteListItem
```

Comment:
I made the change to pre-allocate.  TBH, I stopped pre-allocating except for very large arrays because IMO it is too easy to do `make([]string, len(something))` instead of `make([]string, 0, len(something))`.  These are very different results, especially when using `append` afterward.

Reply:
Indeed. I've never really thought about the danger there. My suggestion is only given to make sure we are consistent so that when it really makes a difference we end up making the right choice. Obviously in this specific case both solutions should work equally well.